src/Mvc/Mvc.RazorPages/src/Infrastructure/DefaultPageLoader.cs

```diff
@@ -41,16 +47,31 @@ internal class DefaultPageLoader : IPageLoader
 
         private IViewCompiler Compiler => _viewCompilerProvider.GetCompiler();
 
-        public CompiledPageActionDescriptor Load(PageActionDescriptor actionDescriptor)
+        public async override ValueTask<CompiledPageActionDescriptor> LoadAsync(PageActionDescriptor actionDescriptor)
         {
             if (actionDescriptor == null)
             {
                 throw new ArgumentNullException(nameof(actionDescriptor));
             }
 
-            var compileTask = Compiler.CompileAsync(actionDescriptor.RelativePath);
-            var viewDescriptor = compileTask.GetAwaiter().GetResult();
+            if (_memoryCache.TryGetValue(actionDescriptor, out CompiledPageActionDescriptor compiledDescriptor))
+            {
+                return compiledDescriptor;
+            }
 
+            var viewDescriptor = await Compiler.CompileAsync(actionDescriptor.RelativePath);
```

Comment:
perf nit: Move this async stuff into its own method and only call when page is not in cache. Then LoadAsync can get from cache without an async state machine.